ivy/functional/frontends/tensorflow/general_functions.py

```diff
@@ -83,6 +84,35 @@ def fill(dims, value, name=None):
     return ivy.full(dims, value)
 
 
+@to_ivy_arrays_and_back
+def foldl(
+    fn,
+    elems,
+    initializer=None,
+    parallel_iterations=10,
+    back_prop=True,
+    swap_memory=False,
+    name=None,
+):
+    try:
+        ivy.utils.assertions.check_isinstance(elems, ivy.Array)
+    except ivy.utils.exceptions.IvyException:
+        ivy.utils.exceptions.IvyException("elems must be an iterable object")
+    if not callable(fn):
+        ivy.utils.exceptions.IvyException(f"{fn.__name__} must be a callable function")
+
+    if initializer is None and elems.shape[0] > 0:
+        result = functools.reduce(fn, elems)
+        if isinstance(result, (int, float)):
+            return ivy.array([result])
+        else:
+            return result
+    elif initializer is not None:
+        return functools.reduce(fn, elems, initializer)
+    else:
+        return elems
```

Comment:
Maybe we can use `ivy.reduce` for the whole implementation after checking for valid input shape? It already used `functools.reduce` inside and got better handling of scalars, instead of trying to reinvent the wheel here :pray:

Reply:
Shifting the implementation to use `ivy.reduce()` is a bit convoluted. First off, the initializer argument is not optional. Regarding the axes arg I believe it should just be kept at the default. Given the current implementation as you see it, replacing all the calls to `functools.reduce` with `ivy.reduce`, the test case for numpy passes, however the other test cases fail with a rather long stack trace. I'll share parts of the TensorFlow test stack trace below: 
```python
args = (ivy.array([]), ivy.array(-1), <function <lambda> at 0x7fa68c099d30>), kwargs = {} #these are the inputs to the test case
operand = ivy.array([]), init_value = ivy.array(-1), computation = <function <lambda> at 0x7fa68c099d30>

    @handle_exceptions
    @handle_nestable
    @inputs_to_ivy_arrays
    def reduce(
        operand: Union[ivy.Array, ivy.NativeArray],
        init_value: Union[int, float],
        computation: Callable,
        /,
        *,
        axes: Union[int, Sequence[int]] = 0,
        keepdims: bool = False,
    ) -> ivy.Array:
        """
        Reduces the input array's dimensions by applying a function along one or more axes.

        Parameters
        ----------
        operand
            The array to act on.
        init_value
            The value with which to start the reduction.
        computation
            The reduction function.
        axes
            The dimensions along which the reduction is performed.
        keepdims
            If this is set to True, the axes which are reduced are left in the result as
            dimensions with size one.

        Returns
        -------
        ret
            The reduced array.

        Examples
        --------
        >>> x = ivy.array([[1, 2, 3], [4, 5, 6]])
        >>> ivy.reduce(x, 0, ivy.add, 0)
        ivy.array([6, 15])

        >>> x = ivy.array([[1, 2, 3], [4, 5, 6]])
        >>> ivy.reduce(x, 0, ivy.add, 1)
        ivy.array([5, 7, 9])
        """
        axes = (axes,) if isinstance(axes, int) else axes
        axes = [a + operand.ndim if a < 0 else a for a in axes]
        axes = sorted(axes, reverse=True)
        init_value = ivy.array(init_value)
        op_dtype = operand.dtype
>       computation = _correct_ivy_callable(computation)

ivy/ivy/functional/ivy/experimental/general.py:72:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

func = <function <lambda> at 0x7fa68c099d30>

    def _correct_ivy_callable(func):
        # get the current backend of the given ivy callable
        if ivy.nested_any(
            func,
            lambda x: hasattr(x, "__module__")
            and x.__module__.startswith("ivy")
            and not x.__module__.startswith("ivy.functional.frontends"),
        ):
>           return ivy.__dict__[func.__name__]
E           KeyError: '<lambda>'

ivy/ivy/functional/ivy/experimental/general.py:19: KeyError

During handling of the above exception, another exception occurred:

    @handle_frontend_test(
>       fn_tree="tensorflow.foldl",
        fn=st.sampled_from(
            [
                lambda a, b: a + b,
                lambda a, b: a - b,
                lambda a, b: a * b,
            ],
        ),
        initializer=st.one_of(st.none(), st.floats(min_value=-1000, max_value=1000)),
        dtype_and_values=helpers.dtype_and_values(
            available_dtypes=helpers.get_dtypes("float", full=False),
            min_value=-1000,
            max_value=1000,
            max_dim_size=10,
            max_num_dims=4,
            min_dim_size=1,
        ),
    )

ivy/ivy_tests/test_ivy/test_frontends/test_tensorflow/test_general_functions.py:272:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
ivy/ivy_tests/test_ivy/test_frontends/test_tensorflow/test_general_functions.py:301: in test_tensorflow_foldl
    helpers.test_frontend_function(
ivy/ivy_tests/test_ivy/helpers/function_testing.py:608: in test_frontend_function
    ret = get_frontend_ret(
ivy/ivy_tests/test_ivy/helpers/function_testing.py:1774: in get_frontend_ret
    ret = frontend_fn(*args, **kwargs)
ivy/ivy/functional/frontends/tensorflow/func_wrapper.py:135: in _outputs_to_frontend_arrays_tf
    ret = fn(*args, **kwargs)
ivy/ivy/functional/frontends/tensorflow/func_wrapper.py:109: in _inputs_to_ivy_arrays_tf
    return fn(*ivy_args, **ivy_kwargs)
ivy/ivy/functional/frontends/tensorflow/general_functions.py:107: in foldl
    return ivy.reduce(elems[1:], elems[0], fn)

E           ivy.utils.exceptions.IvyBackendException: tensorflow: reduce:   File "/ivy/ivy/ivy/utils/exceptions.py", line 210, in _handle_exceptions
E               return fn(*args, **kwargs)
E             File "/ivy/ivy/ivy/func_wrapper.py", line 972, in _handle_nestable
E               return fn(*args, **kwargs)
E             File "/ivy/ivy/ivy/func_wrapper.py", line 482, in _inputs_to_ivy_arrays
E               return fn(*ivy_args, **ivy_kwargs)
E             File "/ivy/ivy/ivy/functional/ivy/experimental/general.py", line 72, in reduce
E               computation = _correct_ivy_callable(computation)
E             File "/ivy/ivy/ivy/functional/ivy/experimental/general.py", line 19, in _correct_ivy_callable
E               return ivy.__dict__[func.__name__]
```
I will continue to look into using `ivy.reduce()` but as of now using `functools.reduce` works fine for the functions being tested for. Perhaps trying to mirror the `tf.foldl` implementation more accurately and use a `ivy.while_loop` would be simpler.

Reply:
Ah the use of `ivy.atleast_1d` is nice. Yea it looks pretty simple now, I just thought maybe `ivy.reduce` could be used, but good job :+1:

Reply:
Got it, however this does lead to a discrepancy between ivy's implementation and tensorflow's, as tensorflow will not accept a scalar while ours will.

Reply:
@FourFortyFour Eh what? So what was the point of `atleast_1d`?

Reply:
I will change it right now, it was because I confused `min_dim_size` and `min_num_dims` in the `dtype_and_values` helper (your question made me reevaluate everything 😅), so when the test function would run, testing `tf.foldl()` would always fail in the `tf.TensorShape`'s `__getitem__` method. It would sometimes be passed scalars, so I added the conversion `ivy.atleast_1d()` to get around that... I thought there was something weird going. Will push fixes right now, thank you for revealing this.